### PR TITLE
Improve test suite focusing on the DAG's behavior

### PR DESCRIPTION
**Pull request by ggalmazor**
This PR improves the test suite by adding new tests and by providing a new DSL to define forms directly in tests.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The only change that could affect users is a small change in `TreeReference.contextualize()` that could be considered fixing of a harmless formal bug. See https://github.com/opendatakit/javarosa/commit/11262ae159d8a431333b040203b414ced232f7f4 for more context about this.

### CODE REVIEW

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 1]
```diff
@@ -385,6 +402,54 @@ public void deleteRepeatGroupWithCalculationsTimingTest() throws Exception {
         logger.info("Deletion of {} repeats took {}", numberOfRepeats, duration.toString());
     }
 
+    @Test
+    public void verify_relation_between_calculate_expressions_and_relevancy_conditions() throws IOException {
+        Scenario scenario = Scenario.init("Interdependencies test", html(
+            head(
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: The main reasons are that there are other init overloads taking strings already and that would require to complicate the Scenario API with other "init" variants, and that passing an `XFormsElement` instance keeps our options open for future implementations where we could decide other strategies to set everything up.

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 2]
```diff
@@ -385,6 +402,54 @@ public void deleteRepeatGroupWithCalculationsTimingTest() throws Exception {
...
+                title("Interdependencies test"),
+                model(
+                    mainInstance(
+                        t("data id=\"interdependencies-test\"",
+                            t("number1"),
+                            t("continue"),
+                            t("group", t("number1_x2"), t("number1_x2_x2"), t("number2")),
+                            t("meta", t("instanceID"))
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: I agree! I thought I removed them in another commit. I'll make sure all of them are removed.

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 3]
```diff
@@ -385,6 +402,54 @@ public void deleteRepeatGroupWithCalculationsTimingTest() throws Exception {
         logger.info("Deletion of {} repeats took {}", numberOfRepeats, duration.toString());
     }
 
+    @Test
+    public void verify_relation_between_calculate_expressions_and_relevancy_conditions() throws IOException {
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: Agreed. I think we should create an issue too. This feels important.

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 4]
```diff
@@ -450,6 +452,80 @@ public void verify_relation_between_calculate_expressions_and_relevancy_conditio
...
     }
 
+    /**
+     * Ignored because the assertions about non-null next-numbers will fail because our DAG
+     * doesn't evaluate calculations in repeat instances that are previous siblings to the
+     * one that has changed.
+     */
+    @Test
+    @Ignore
+    public void calculate_expressions_should_be_evaluated_on_previous_repeat_siblings() throws IOException {
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: I think this is not only related to deleting repeats, but also to changing the value of a field inside a repeat group. I'll add the comment, though.

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 5]
```diff
@@ -450,6 +452,80 @@ public void verify_relation_between_calculate_expressions_and_relevancy_conditio
...
+        scenario.next();
+        scenario.answer(22);
+
+        assertThat(scenario.answerOf("/data/group[0]/prev-number"), is(nullValue()));
+        assertThat(scenario.answerOf("/data/group[1]/prev-number"), is(intAnswer(11)));
+
+        assertThat(scenario.answerOf("/data/group[0]/number"), is(intAnswer(11)));
+        assertThat(scenario.answerOf("/data/group[1]/number"), is(intAnswer(22)));
+
+        assertThat(scenario.answerOf("/data/group[0]/next-number"), is(intAnswer(22)));
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: Agreed. I considered adding the comment you are suggesting but wanted to see if this attracted your attention during review ;)

I'll update the doc block and add comments on the failing assertions.

File: src/test/java/org/javarosa/core/model/instance/TreeReferenceIsAncestorOfTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            {"/foo[-1]/bar[2] is parent of /foo/bar[2]/baz", "/foo[-1]/bar[2]", "/foo/bar[2]/baz", IRRELEVANT, true},
+            {"/foo[-1]/bar[2] is parent of /foo[-1]/bar[2]/baz", "/foo[-1]/bar[2]", "/foo[-1]/bar[2]/baz", IRRELEVANT, true},
+            {"/foo[-1]/bar[2] is parent of /foo[0]/bar[2]/baz", "/foo[-1]/bar[2]", "/foo[0]/bar[2]/baz", IRRELEVANT, true},
+            {"/foo[-1]/bar[2] is parent of /foo[2]/bar[2]/baz", "/foo[-1]/bar[2]", "/foo[2]/bar[2]/baz", IRRELEVANT, true},
+            {"/foo[2]/bar[2] is parent of /foo[3]/bar[2]/baz", "/foo[2]/bar[2]", "/foo[3]/bar[2]/baz", IRRELEVANT, false},
+        });
+    }
+
+    @Test
+    public void parent_works_as_expected() {
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/instance/TreeReferenceIsAncestorOfTest.java]: Oops, good catch! I'll change names here too.

File: src/main/java/org/javarosa/core/model/instance/TreeReference.java [hunk 7]
```diff
@@ -328,6 +328,7 @@ public TreeReference contextualize (TreeReference contextRef) {
                 //may have been done.
                 if(newRef.getPredicate(i) == null) {
                     newRef.setMultiplicity(i, contextRef.getMultiplicity(i));
+                    newRef.addPredicate(i, contextRef.getPredicate(i));
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/main/java/org/javarosa/core/model/instance/TreeReference.java]: Yeah, this is one of the things that made understanding JR super hard for me... I wish there would be only one concept, be it multiplicity or predicate, and we wouldn't have special values in multiplicity like `-1`, `-4`, etc. meaning different things...

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java [hunk 8]
```diff
@@ -425,6 +425,51 @@ public void calculate_expressions_should_be_evaluated_on_previous_repeat_sibling
         assertThat(scenario.answerOf("/data/group[2]/next-number"), is(nullValue()));
     }
 
+    /**
+     * Ignored because the count() function inside the predicate of the result_2 calculate
+     * expression isn't evaled correctly, as opposed to the predicate of the result_1 calculate
+     * that uses an interim field as a proxy for the same computation
+     */
```
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: Agreed!

**Comment by ggalmazor** [comment]
So, regarding the DSL, here's my mindset around it:
- It's good that I need to know how XForms work to use the DSL. I woudn't want a DSL that prevents me from learning the XForms syntax. 

  A great example of DSLs that work this way is [jOOQ for SQL](https://www.jooq.org)
- Right now, this DSL is super dumb, but it can be improved by providing static type checking, for example, which would make using it easier for new users e.g. the `model` factory takes in `instance` and `bind` objects only.
- I'd argue that form parts are more reusable if you define them using the DSL instead of using just strings.

  A common scenario in our test suite is that we have different test methods using slight variations of the same form. This could be improved by abstracting away what's common and leaving only the moving parts in the test method. All this should be easier to do with factories that use the DSL instead of strings.
- Overall, having a DSL makes it easier for us to implement more sophisticated tools in the future

File: src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ggalmazor) [reply src/test/java/org/javarosa/core/model/Safe2014DagImplTest.java]: Created https://github.com/opendatakit/javarosa/issues/516

**Comment by seadowg** [comment]
I might have the wrong idea of what's happening here but should `isParent` be deprecated (and just call through to `isAncestor`) here rather than just straight renamed? It looks like this causes problems for Collect (#3536) and I'm wondering if any other clients might be hit.

**Comment by ggalmazor** [comment]
Yes, @seadowg, it was a mistake to rename it and not provide an alias, sorry! https://github.com/opendatakit/javarosa/issues/519 already tracks this and @grzesiek2010 is working on a fix.

**Comment by seadowg** [comment]
@ggalmazor ach sorry I totally missed that!

**Comment by ggalmazor** [comment]
Don't worry ;)

**Pull request merged**